File: models/Gravity/gravity_highprecision.c
#include <math.h>
#include "gravity_highprecision.h"
#include "constants.h"


#if   defined(GRAVITY_EARTH_EGM96)
    #include "EGM96_parameters.h"

#elif defined(GRAVITY_EARTH_EGM2008)
    #include "EGM2008_parameters.h"

#elif defined(GRAVITY_MOON_LP100K)
    #include "LP100K_parameters.h"

#elif defined(GRAVITY_MOON_LP165P)
    #include "LP165P_parameters.h"

#elif defined(GRAVITY_MARS_GMM2B)
    #include "GMM2B_parameters.h"

#else
    #error Missing gravity model definition.
#endif


/*
 * References:
 * [1] "Fast Gravity, Gravity Partials, Normalized Gravity, Gravity Gradient
 *      Torque and Magnetic Field: Derivation, Code and Data". 1993.
 *      Robert G. Gottlieb,  McDonnell Douglas Space Systems, Houston
 *      Devision, Houston Texas. Prepared for Lyndon B. Johnson Space Center,
 *      contract NAS9-17885.
*/


static double nn[GRAVITY_COEFF_MAXDEGREE] = 0.0,
              mn[GRAVITY_COEFF_MAXDEGREE] = 0.0;


/* Fully normalized associated Legendre polynomials */
static
void
ComputeLegendrePolys(double phi,
                     unsigned int nmax,
                     double (*Pnm[GRAVITY_COEFF_MAXDEGREE+3u][GRAVITY_COEFF_MAXDEGREE+3u]))
{
    /* Computation based on equations 7-9 and 7-10 from [1] */

    static unsigned int Pnm_initialized = 0u;

    static double  xi[GRAVITY_COEFF_MAXDEGREE][GRAVITY_COEFF_MAXDEGREE] = 0.0,
                  eta[GRAVITY_COEFF_MAXDEGREE][GRAVITY_COEFF_MAXDEGREE] = 0.0;

    unsigned int n,
                 m;

    double cphi = cos(PI_OVER_2 - phi),
           sphi = sin(PI_OVER_2 - phi);

    if (fabs(cphi) <= EPS) cphi = 0.0;
    if (fabs(sphi) <= EPS) sphi = 0.0;

    if (!Pnm_initialized)
    {
        /* TODO: implement eqs. 7-10 and 7-12 */
    }



    (*Pnm)[0][0] = 1.0;
    (*Pnm)[1][0] = SQUAREROOT3 * cphi;
    (*Pnm)[1][1] = SQUAREROOT3 * sphi;

    for (n=2u; n<nmax+3u; ++n)
    {
        unsigned int nm1 = n - 1u;

        double nm1d  = nn[n] - 1.0,
               n2    = nn[n] * 2.0,
               sn2   = sqrt(n2+0.0),  sn2p1 = sqrt(n2+1.0),
               sn2m3 = sqrt(n2-3.0),  sn2m1 = sqrt(n2-1.0);

        (*Pnm)[n][0] = sn2p1/nn[n] *
                         (sn2m1*cphi*(*Pnm)[nm1][0u] - nm1d/sn2m3 * (*Pnm)[n-2u][0u]);

        for (m=1u; m<n+1u; ++m) {
            (*Pnm)[n][m] = sn2p1 / ( sqrt(nn[n]+mm[m]) * sqrt(nn[n]-mm[m]) ) *
                             (sn2m1*cphi*(*Pnm)[nm1][m] - sqrt(nm1d+mm[m]) * sqrt(nm1d-mm[m]) * (*Pnm)[n-2u][m]/sn2m3);
        }

        (*Pnm)[n][n] = sn2p1/sn2 * sphi*(*Pnm)[nm1][nm1];

    }

}


void
gravity_highprecision(const double pos_ECEF[3],
                      double JD,
                      unsigned int nmax,
                      int compute_gradient,   double (*gradient)[3],
                      int compute_hessian,    double (*hessian)[3][3])
{
    static unsigned int grav_initialized = 0u;

    double Pnm[GRAVITY_COEFF_MAXDEGREE+3u][GRAVITY_COEFF_MAXDEGREE+3u] = 0.0,
           cmlambda[GRAVITY_COEFF_MAXDEGREE+1u] = 0.0,
           smlambda[GRAVITY_COEFF_MAXDEGREE+1u] = 0.0;

    unsigned int i,
                 n,
                 m;

    double deltaT                 = (JD - Gravity_reference_epoch)/DAYS_PER_YEAR,
           dUdr,      dUdr_n      = 1.0,
           dUdphi,    dUdphi_n    = 0.0,
           dUdlambda, dUdlambda_n = 0.0;

    double x = pos_ECEF[0],
           y = pos_ECEF[1],
           z = pos_ECEF[2],

           r      = sqrt (x*x + y*y + z*z),
           phi    = asin (z/r),
           lambda = atan2(y,x),

           slambda = sin(lambda),
           clambda = cos(lambda),

           rm1 = 1.0/r,
           rm2 = rm1*rm1,

           roR   = Gravity_Re*rm1,  xy2  = x*x + y*y,
           roR_n = roR,             nxy  = sqrt(xy2),
           oxy2  = 1/xy2,           onxy = 1/nxy;

    /* Casting the integers in a loop is quite expensive; so cast them once at
     * program start into memory, and reuse them */
    if (!grav_initialized)
    {
        for (n=0u; n<GRAVITY_COEFF_MAXDEGREE; ++n) {
            nn[n] = (double)n;
            mm[n] = nn[n];
        }
        grav_initialized = 1u;
    }

    /* Initialize output */
    for (i=0u; i<3u; ++i) (*gradient)[i] = 0.0;
    for (i=0u; i<9u; ++i) (*hessian)[i]  = 0.0;

    /* Fully normalized associated Legendre polynomials */
    ComputeLegendrePolys(phi, nmax, &Pnm);


    /* Compute partial derivatives of potential
     * ================================================================================ */

    /* pre-compute trig factors (recursion relations) */
    smlambda[0] = slambda;  smlambda[1] = 2.0*clambda*slambda;
    cmlambda[0] = clambda;  cmlambda[1] = 2.0*clambda*clambda - 1.0;

    for (m=2u; m<=nmax; ++m)
    {
        smlambda[m] = 2.0*clambda*smlambda[m-1u] - smlambda[m-2u];
        cmlambda[m] = 2.0*clambda*cmlambda[m-1u] - cmlambda[m-2u];
    }

    /* Compute double sum */
    for (n=2u; n<=nmax; ++n)
    {
        double dUdr_m      = 0.0,
               dUdphi_m    = 0.0,
               dUdlambda_m = 0.0;

        /* summation over m */
        for (m=0u; m<=n; ++m)
        {
            double fac1, Cnm = Gravity_Cnm[n][m],
                   fac2, Snm = Gravity_Snm[n][m];

            /* Take into account coefficient drifts */
            if (n < GRAVITY_DRIFT_MAXDEGREE && m < GRAVITY_DRIFT_MAXDEGREE) {
                Cnm += DeltaT * Gravity_Cnm_dot[n][m];
                Snm += DeltaT * Gravity_Snm_dot[n][m];
            }

            fac1 = Cnm * cmlambda[m] + Snm * smlambda[m],
            fac2 = Snm * cmlambda[m] - Cnm * smlambda[m];

            dUdr_m      += Pnm[n][m] * fac1;
            dUdphi_m    += (Pnm[n][m+1u] - z*onxy * mm[m]*Pnm[n][m]) * fac1;
            dUdlambda_m += mm[m]*Pnm[n][m] * fac2;
        }

        /* summation over n */
        roR_n       *= roR;
        dUdr_n      += roR_n * dUdr_m * (nn[n] + 1.0);
        dUdphi_n    += roR_n * dUdphi_m;
        dUdlambda_n += roR_n * dUdlambda_m;

    }

    /* Acceleration in spherical coordinates */
    dUdr      = -Gravity_GM*rm2 * dUdr_n;
    dUdphi    =  Gravity_GM*rm1 * dUdphi_n;
    dUdlambda =  Gravity_GM*rm1 * dUdlambda_n;

    /**/
    if (compute_gradient)
    {
        /* Special case for positions near the poles */
        if (fabs(atan2(z,nxy)) == PI_OVER_2)
        {
            (*gradient)[0] = 0.0;
            (*gradient)[1] = 0.0;
            (*gradient)[2] = z*dUdr*rm1;
        }

        /* All other locations: */
        else
        {
            /* Convert back to Planet centered/planet fixed */
            double f = dUdr*rm1,
                   g = f - z*dUdphi*rm2*onxy,
                   h = dUdlambda*oxy2;

            (*gradient)[0] = g*x - h*y;
            (*gradient)[1] = g*y + h*x;
            (*gradient)[2] = f*z + nxy*dUdphi*rm2;
        }
    }

    /* Compute second derivatives (Hessian) */
    if (compute_hessian)
    {
        /* TODO: eq 5-15, 8-30*/
    }
}

File: models/Gravity/gravity_highprecision.m
function [acceleration, Jacobian] = gravity_highprecision(p, maxdeg, model) %#eml



%  GRAVITYSPHERICALHARMONIC Implement a spherical harmonic representation
%   of planetary gravity.
%   [GX GY GZ] = GRAVITYSPHERICALHARMONIC( P ) implements the mathematical
%   representation of spherical harmonic planetary gravity based on
%   planetary gravitational potential. Using P, a M-by-3 array of
%   Planet-Centered Planet-Fixed coordinates, GX, GY and GZ, arrays of M
%   gravity values in the x-axis, y-axis and z-axis of the Planet-Centered
%   Planet-Fixed coordinates are calculated for planet using 120th degree
%   and order spherical coefficients for EGM2008 by default.
%
%   Alternate formats for calling zonal harmonic gravity are:
%   [GX GY GZ] = GRAVITYSPHERICALHARMONIC( P, DEGREE )
%   [GX GY GZ] = GRAVITYSPHERICALHARMONIC( P, MODEL )
%   [GX GY GZ] = GRAVITYSPHERICALHARMONIC( P, MODEL, DEGREE )
%   [GX GY GZ] = GRAVITYSPHERICALHARMONIC( P, MODEL, DEGREE, ACTION )
%   [GX GY GZ] = GRAVITYSPHERICALHARMONIC( P, 'Custom', DEGREE, {DATAFILE DFREADER}, ACTION )
%
%   Inputs for spherical harmonic gravity are:
%   P        :a M-by-3 array of Planet-Centered Planet-Fixed coordinates in
%            meters where the z-axis is positive towards the North Pole. For
%            Earth this would be ECEF coordinates.
%   MODEL    :a string specifying the planetary model:
%            'EGM2008' (Earth), 'EGM96' (Earth), 'LP100K' (Moon), 'LP165P'
%            (Moon), 'GMM2B' (Mars), or 'Custom'.  The default is 'EGM2008'.
%   DEGREE   :a scalar value specifying the degree and order of the
%            harmonic gravity model. For 'EGM2008', the maximum degree and
%            order is 2159 and the default degree and order is 120.   For
%            'EGM96', the maximum degree and order is 360 and the default
%            degree and order is 70.  For 'LP100K', the maximum degree and
%            order is 100 and the default degree and order is 60.  For
%            'LP165P', the maximum degree and order is 165 and the default
%            degree and order is 60.  For 'GMM2B', the maximum degree and
%            order is 80 and the default degree and order is 60.  For
%            'Custom', the default degree and order is the maximum degree.
%   DATAFILE :a file containing the planetary gravitational parameter,
%            planet equatorial radius, maximum degree, and normalized
%            spherical harmonic coefficient matrices.
%   DFREADER :a function handle to an MATLAB(R) function which reads
%            DATAFILE.  The MATLAB function must output planetary
%            gravitational parameter in meters cubed per second squared,
%            planet equatorial radius in meters, maximum degree, and the
%            normalized spherical harmonic coefficient matrices, C and S.
%   ACTION   :a string to determine action for out of range input. Specify
%            if out of range input invokes a 'Warning', 'Error', or no
%            action ('None'). The default is 'Warning'.
%
%   Output calculated for the spherical harmonic gravity includes:
%   GX     :an array of M gravity values in the x-axis of the
%          Planet-Centered Planet-Fixed coordinates in meters per second
%          squared.
%   GY     :an array of M gravity values in the y-axis of the
%          Planet-Centered Planet-Fixed coordinates in meters per second
%          squared.
%   GZ     :an array of M gravity values in the z-axis of the
%          Planet-Centered Planet-Fixed coordinates in meters per second
%          squared.
%
%   Limitations:
%
%   This function has the limitations of excluding the centrifugal effects
%   of planetary rotation, and the effects of a precessing reference frame.
%
%   Spherical harmonic gravity model is valid for radial positions greater
%   than the planet equatorial radius.  Using it near or at the planetary
%   surface can probably be done with negligible error.  The spherical
%   harmonic gravity model is not valid for radial positions less than
%   planetary surface.
%
%   Examples:
%
%   Calculate the gravity in the x-axis at the equator on the surface of
%   Earth, using the 120 degree model of EGM2008 with warning actions:
%       gx = gravitysphericalharmonic( [-6378.1363e3 0 0] )
%
%   Calculate the gravity at 25000 meters over the south pole of Earth using
%   the 70 degree model of EGM96 with error actions:
%       [gx, gy, gz] = gravitysphericalharmonic( [0 0 -6381.751e3], 'EGM96', 'Error' )
%
%   Calculate the gravity at 15000 meters over the equator and 11000 meters
%   over the north pole using a 30th order GMM2B Mars model with warning
%   actions:
%       p  = [2412.648e3 -2412.648e3 0; 0 0 3376.2e3]
%       [gx, gy, gz] = gravitysphericalharmonic( p, 'GMM2B', 30, 'Warning' )
%
%   Calculate the gravity at 15000 meters over the equator and 11000 meters
%   over the north pole using a 60th degree custom planetary model with no
%   actions:
%       p       = [2412.648e3 -2412.648e3 0; 0 0 3376e3]
%       [gx, gy, gz] = gravitysphericalharmonic( p, 'custom', 60, ...
%                       {'GMM2BC80_SHA.txt' @astReadSHAFile}, 'None' )
%
%   See also GRAVITYWGS84, GRAVITYCENTRIFUGAL, GRAVITYZONAL, GEOIDEGM96

%   References:
%   [1] Vallado, D. A., "Fundamentals of Astrodynamics and Applications",
%       McGraw-Hill, New York, 1997.
%   [2] NIMA TR8350.2: "Department of Defense World Geodetic System 1984,
%       Its Definition and Relationship with Local Geodetic Systems."
%   [3] Konopliv, A. S., S. W. Asmar, E. Carranza, W. L. Sjogen, D. N.
%       Yuan., "Recent Gravity Models as a Result of the Lunar Prospector
%       Mission", Icarus, Vol. 150, no. 1, pp 1?18, 2001.
%   [4] Lemoine, F. G., D. E. Smith, D.D. Rowlands, M.T. Zuber, G. A.
%       Neumann, and D. S. Chinn, "An improved solution of the gravity
%       field of Mars (GMM-2B) from Mars Global Surveyor", J. Geophys. Res.,
%       Vol. 106, No. E10, pp 23359-23376, October 25, 2001.
%   [5] Kenyon S., J. Factor, N. Pavlis, and S. Holmes, "Towards the Next
%       Earth Gravitational Model", Society of Exploration Geophysicists
%       77th Annual Meeting, San Antonio, Texas, September 23-28, 2007.
%   [6] Pavlis, N.K., S.A. Holmes, S.C. Kenyon, and J.K. Factor, "An Earth
%       Gravitational Model to Degree 2160: EGM2008", presented at the 2008
%       General Assembly of the European Geosciences Union, Vienna,
%       Austria, April 13-18, 2008.
%   [7] Grueber, T., and A. Kohl, "Validation of the EGM2008 Gravity Field
%       with GPS-Leveling and Oceanographic Analyses", presented at the IAG
%       International Symposium on Gravity, Geoid & Earth Observation 2008,
%       Chania, Greece, June 23-27, 2008.

    %% Initialize

    persistent GM Re degree C S %#ok
    persistent prevmodel
    if isempty(prevmodel)
        prevmodel = 0; end 

    switch model 

        % [GM, Re, degree, C, S]
        case 1%'egm2008'
            % Earth            
            if prevmodel~=1
                load('aeroegm2008.mat')                 
                prevmodel = 1;
            end
            
        case {0 2}%'egm96'
            % Earth            
            if prevmodel~=2
                load('aeroegm96.mat') 
                prevmodel = 2;
            end
          
            %{
        case 3%'lp100k'
            % Moon            
            load('aerolp100k.mat') 
            
        case 4%'lp165p'
            % Moon            
            load('aerolp165p.mat') 
            
        case 5%'gmm2b'
            % Mars            
            load('aerogmm2b.mat') 
            %}
            
    end

    % convert to spherical coordinates
    r      = norm(p);
    phi    = asin (p(3)/r);
    lambda = atan2(p(2),p(1));

    % Check if we're flying underground
    if r <= Re
        % TODO
    end
    
    % Often-used values
    rm1 = 1/r;      pio2  = pi/2;             
    rm2 = rm1*rm1;  sqrt3 = sqrt(3);

    
    %% Fully normalized associated Legendre polynomials 
    
    % Computed via recursion relations
    
    P = zeros(maxdeg+3, maxdeg+3);
    
    cphi = cos(pio2 - phi);    cphi(abs(cphi) <= eps) = 0;
    sphi = sin(pio2 - phi);    sphi(abs(sphi) <= eps) = 0;
    
    P(1,1) = 1;            
    P(2,1) = sqrt3 * cphi; 
    P(2,2) = sqrt3 * sphi; 

    for n = 2 : maxdeg+2
        
        k  = n + 1;
        n2 = 2*n;
        
        sn2   = sqrt(n2+0);  sn2p1 = sqrt(n2+1); 
        sn2m3 = sqrt(n2-3);  sn2m1 = sqrt(n2-1);
                
        P(k,1) = sqrt(n2+1)/n * ...
                 (sn2m1*cphi*P(k-1,1) - (n-1)/sn2m3 * P(k-2,1));             
        P(k,k) = sn2p1/sn2 * sphi*P(k-1,k-1);  
        
        for m = 1 : n-1
            q = m + 1;            
            P(k,q) = sn2p1 / (sqrt(n+m)*sqrt(n-m)) * ...
                     (sn2m1*cphi*P(k-1,q) - sqrt(n+m-1)*sqrt(n-m-1)/sn2m3*P(k-2,q));
        end
        
    end
    
    
    %% Compute partial derivatives of potential 
    
    % pre-compute trig factors (recursion relations)
    slambda  = sin(lambda);  
    clambda  = cos(lambda);  
    
    smlambda = [slambda  2*clambda*slambda     zeros(1, maxdeg-1)];    
    cmlambda = [clambda  2*clambda*clambda-1   zeros(1, maxdeg-1)];
    
    for m = 3:maxdeg+1
        smlambda(m) = 2*clambda*smlambda(m-1) - smlambda(m-2);
        cmlambda(m) = 2*clambda*cmlambda(m-1) - cmlambda(m-2);
    end
    
    % Other often-used values    
    roR  = Re*rm1;             roR_n = roR;
    xy2  = p(1)^2 + p(2)^2;    nxy   = sqrt(xy2);
    oxy2 = 1/xy2;              onxy  = 1/nxy;

    % Summations in radial coordinates    
    dUdr_n      = 1;
    dUdphi_n    = 0;
    dUdlambda_n = 0;
    for n = 2:maxdeg
        
        k     = n+1;
        roR_n = roR_n * roR;
        
        % (summation over m)
        fac1        = ( C(k,1:k).*cmlambda(1:k) + S(k,1:k).*smlambda(1:k) ).';
        fac2        = ( S(k,1:k).*cmlambda(1:k) - C(k,1:k).*smlambda(1:k) ).';
        dUdr_m      = P(k,1:k) * fac1;
        dUdphi_m    = (P(k,2:k+1) - p(3)*onxy * (0:n).*P(k,1:k)) * fac1;
        dUdlambda_m = (0:n).*P(k,1:k) * fac2;
        
        % (summation over n)
        dUdr_n      = dUdr_n      + k*roR_n*dUdr_m;     
        dUdphi_n    = dUdphi_n    +   roR_n*dUdphi_m;   
        dUdlambda_n = dUdlambda_n +   roR_n*dUdlambda_m;
        
    end

    % Acceleration in spherical coordinates
    dUdr      = -GM*rm2 * dUdr_n;
    dUdphi    =  GM*rm1 * dUdphi_n;
    dUdlambda =  GM*rm1 * dUdlambda_n;
    
    % Special case for positions near the poles
    if abs(atan2(p(3),nxy)) == pio2
        
        acceleration = [
            0
            0
            p(3)*dUdr*rm1;
            ];
    
    % All other locations:
    else   
        % Convert back to ECEF
        f = dUdr*rm1;
        g = f - p(3)*dUdphi*rm2*onxy;
        h = dUdlambda*oxy2;
        
        acceleration = [
            g*p(1) - h*p(2)
            g*p(2) + h*p(1)
            f*p(3) + nxy*dUdphi*rm2
            ];
    end

    
    %% Compute partial derivatives (Jacobian) 

end



